lib/auth/touchid/api_darwin.go

```diff
@@ -0,0 +1,210 @@
+		return "", ""
+	}
+	rpID := label[0:idx]
+	user := label[idx+1:]
+	return rpID, user
+}
+
+func (*touchIDImpl) Authenticate(credentialID string, digest []byte) ([]byte, error) {
+	var req C.AuthenticateRequest
+	req.app_label = C.CString(credentialID)
+	req.digest = C.CString(string(digest))
+	req.digest_len = C.int(len(digest))
+	defer func() {
+		C.free(unsafe.Pointer(req.app_label))
+		C.free(unsafe.Pointer(req.digest))
+	}()
+
+	var sigOutC, errMsgC *C.char
+	defer func() {
+		C.free(unsafe.Pointer(sigOutC))
+		C.free(unsafe.Pointer(errMsgC))
+	}()
+
+	if res := C.Authenticate(req, &sigOutC, &errMsgC); res != 0 {
+		errMsg := C.GoString(errMsgC)
+		return nil, errors.New(errMsg)
+	}
+
+	sigB64 := C.GoString(sigOutC)
+	return base64.StdEncoding.DecodeString(sigB64)
+}
+
+func (*touchIDImpl) FindCredentials(rpID, user string) ([]CredentialInfo, error) {
+	infos, res := findCredentialsImpl(rpID, user, func(filter C.LabelFilter, infosC **C.CredentialInfo) C.int {
+		return C.FindCredentials(filter, infosC)
+	})
+	if res < 0 {
+		return nil, fmt.Errorf("failed to find credentials: status %d", res)
+	}
+	return infos, nil
+}
+
+func findCredentialsImpl(rpID, user string, find func(C.LabelFilter, **C.CredentialInfo) C.int) ([]CredentialInfo, int) {
+	var filterC C.LabelFilter
+	if user == "" {
+		filterC.kind = C.LABEL_PREFIX
+	}
+	filterC.value = C.CString(makeLabel(rpID, user))
+	defer C.free(unsafe.Pointer(filterC.value))
+
+	var infosC *C.CredentialInfo
+	defer C.free(unsafe.Pointer(infosC))
+
+	res := find(filterC, &infosC)
+	if res < 0 {
+		return nil, int(res)
+	}
+
+	start := unsafe.Pointer(infosC)
+	size := unsafe.Sizeof(C.CredentialInfo{})
+	infos := make([]CredentialInfo, res)
+	for i := 0; i < int(res); i++ {
+		// IMPORTANT: The defer below is used to free the pointers inside infos.
+		// It relies on the fact that we never error out of the function after
+		// this point, otherwise some instances would leak.
+		infoC := (*C.CredentialInfo)(unsafe.Add(start, uintptr(i)*size))
+		defer func() {
+			C.free(unsafe.Pointer(infoC.label))
+			C.free(unsafe.Pointer(infoC.app_label))
+			C.free(unsafe.Pointer(infoC.app_tag))
+			C.free(unsafe.Pointer(infoC.pub_key_b64))
+		}()
+
+		// user@rpid
+		label := C.GoString(infoC.label)
+		rpID, user := splitLabel(label)
+		if rpID == "" || user == "" {
+			log.Debugf("Skipping credential with unexpected label: %q", label)
+			continue
+		}
+
+		// credential ID / UUID
+		credentialID := C.GoString(infoC.app_label)
+
+		// user handle
+		appTag := C.GoString(infoC.app_tag)
+		userHandle, err := base64.RawURLEncoding.DecodeString(appTag)
+		if err != nil {
+			log.Debugf("Skipping credential with unexpected application tag: %q", appTag)
+			continue
+		}
+
+		// ECDSA public key
+		pubKeyB64 := C.GoString(infoC.pub_key_b64)
+		pubKeyRaw, err := base64.StdEncoding.DecodeString(pubKeyB64)
+		if err != nil {
+			log.WithError(err).Warnf("Failed to decode public key for credential %q", credentialID)
```

Comment:
Just making sure, if we fail to decode the public key we still want to return it empty?
https://github.com/gravitational/teleport/pull/12436/files#diff-8527289b15a0ea294e696a93ff2030c016939f5d40e48c081d922a23ffde911dR206

Reply:
Correct, we want to keep going for the defers to run. I'll leave a comment.